sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryRowLevelOperationTable.scala

```diff
@@ -0,0 +1,96 @@
+
+import java.util
+
+import org.apache.spark.sql.connector.distributions.{Distribution, Distributions}
+import org.apache.spark.sql.connector.expressions.{FieldReference, LogicalExpressions, NamedReference, SortDirection, SortOrder, Transform}
+import org.apache.spark.sql.connector.read.{Scan, ScanBuilder}
+import org.apache.spark.sql.connector.write.{BatchWrite, LogicalWriteInfo, RequiresDistributionAndOrdering, RowLevelOperation, RowLevelOperationBuilder, RowLevelOperationInfo, Write, WriteBuilder, WriterCommitMessage}
+import org.apache.spark.sql.connector.write.RowLevelOperation.Command
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+class InMemoryRowLevelOperationTable(
+    name: String,
+    schema: StructType,
+    partitioning: Array[Transform],
+    properties: util.Map[String, String])
+  extends InMemoryTable(name, schema, partitioning, properties) with SupportsRowLevelOperations {
+
+  override def newRowLevelOperationBuilder(
+      info: RowLevelOperationInfo): RowLevelOperationBuilder = {
+    () => PartitionBasedOperation(info.command)
+  }
+
+  case class PartitionBasedOperation(command: Command) extends RowLevelOperation {
+    private final val PARTITION_COLUMN_REF = FieldReference(PartitionKeyColumn.name)
+
+    var configuredScan: InMemoryBatchScan = _
+
+    override def requiredMetadataAttributes(): Array[NamedReference] = {
+      Array(PARTITION_COLUMN_REF)
+    }
+
+    override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder = {
+      new InMemoryScanBuilder(schema) {
+        override def build: Scan = {
+          val scan = super.build()
+          configuredScan = scan.asInstanceOf[InMemoryBatchScan]
+          scan
+        }
+      }
+    }
+
+    override def newWriteBuilder(info: LogicalWriteInfo): WriteBuilder = new WriteBuilder {
+
+      override def build(): Write = new Write with RequiresDistributionAndOrdering {
+        override def requiredDistribution(): Distribution = {
+          Distributions.clustered(Array(PARTITION_COLUMN_REF))
+        }
+
+        override def requiredOrdering(): Array[SortOrder] = {
+          Array[SortOrder](
+            LogicalExpressions.sort(
+              PARTITION_COLUMN_REF,
+              SortDirection.ASCENDING,
+              SortDirection.ASCENDING.defaultNullOrdering())
+          )
+        }
+
+        override def toBatch: BatchWrite = PartitionBasedReplaceData(configuredScan)
+
+        override def description(): String = "InMemoryWrite"
+      }
+    }
+
+    override def description(): String = "InMemoryPartitionReplaceOperation"
+  }
+
+  private case class PartitionBasedReplaceData(scan: InMemoryBatchScan) extends TestBatchWrite {
+
+    override def commit(messages: Array[WriterCommitMessage]): Unit = dataMap.synchronized {
+      val newData = messages.map(_.asInstanceOf[BufferedRows])
+      val readRows = scan.data.flatMap(_.asInstanceOf[BufferedRows].rows)
```

Comment:
This is the place where the writer can obtain unique groups IDs that were read (i.e. files in case of Delta).

Those group IDs can be filtered in two ways:
- static
- runtime

The proposal is that Spark will first should push down whatever possible conditions so that data sources can use static filters to prune partitions and files using metadata. You are right subqueries won't be pushed down but there may be other simple conditions. That gives us a set of files that **_may_** potentially match (still has false positives). In the example you provided, static filtering is not possible since we only have a subquery. Next, Spark should scan the potentially matching files using a separate filtering query to reduce the amount of data to rewrite. That should give us the minimum set of data files to rewrite. I agree with the filtering subquery you provided.

I think we are talking about the same algorithm but I still believe the query you mentioned can be executed dynamically and the unique files can be collected and passed as runtime IN filter.

Let me show how plans can look like for a DELETE statement with an IN subquery.

```
DELETE FROM t WHERE id IN (SELECT * FROM deleted_id)
```

```
== Optimized Logical Plan ==
ReplaceData RelationV2[id#66, dep#67] t
+- Project [id#66, dep#67]
   +- Filter NOT (exists#86 <=> true)
      +- Join ExistenceJoin(exists#86), (id#66 = value#19)
         :- Project [id#66, dep#67]
         :  +- Filter dynamicpruning#85 [_file_name#70]
         :     :  +- Project [_file_name#83]
         :     :     +- Join LeftSemi, (id#81 = value#72)
         :     :        :- Project [id#81, _file_name#83]
         :     :        :  +- RelationV2[id#81, dep#82, _file_name#83] t
         :     :        +- LocalRelation [value#72]
         :     +- RelationV2[id#66, dep#67, _file_name#70] t
         +- LocalRelation [value#19]
```

Just like you said, the filtering subquery would be planned as a scalable left semi join.

Does that make sense? I think we are talking about the same algorithm, just different representation.

Reply:
Yea, we are talking about the same algorithm, just different ways to do interaction between Spark and data sources:
1. Let the data source implicitly obtain the affected "groups" within itself, making the API simpler. (The current proposal)
2. Ask the data source to report "group id" so that Spark can calculate the affected groups by itself. (My new proposal)

The above query plan you provided makes sense, the problem is how to build that subquery filter (the join) without any DS v2 API for the data source to report the "group id"?

Reply:
The current proposal is to leverage existing `SupportsRuntimeFiltering`. Data sources that support row-level operations will need to implement `RowLevelOperation` added in this PR. That one is used to instantiate a scan builder, which means data sources can provide a special scan for row-level operations. Runtime filtering in DS V2 is very flexible and can support metadata columns.

In case of Delta, it should be sufficient to have a scan that would report `_file_name` as a filtering attribute and then Spark will execute a subquery and pass unique file names back into the scan. Since the write will have access to the scan, it will also know the set of files that was scanned and needs to be replaced with new files.

I was trying to keep the row-level API as simple as possible and rely on existing concepts. This approach builds on top of metadata columns and runtime filtering we already support.

Reply:
In case of regular Hive tables, runtime filtering can still use data columns.

Reply:
It is also extremely easy to build the filtering subquery. In case of DELETE, we just need to find matches so we have to take the delete condition and evaluate it against the target table. Spark will take care of rewriting that condition as joins using existing rules.

Reply:
Something like this:

```
// construct a filtering plan with the original scan relation
val matchingRowsPlan = command match {
  case d: DeleteFromTable =>
    Filter(d.condition.get, relation)

  case u: UpdateTable =>
    Filter(u.condition.get, relation)

  case m: MergeIntoTable =>
    Join(relation, m.sourceTable, LeftSemi, Some(m.mergeCondition), JoinHint.NONE)
}
```

Reply:
Do you mean Spark just builds the `Filter` query, runs it and discards the result? The data source `Scan` should find a way to collect the affected "groups" by itself? I think it works, but it's a bit awkward/too implicit...

Reply:
I don't think the result will be discarded. Let me explain how I see this:

- The data source plans input splits for groups that may have matches using parts of the condition that can be converted into data source filters. 
- The data source can cache group IDs or input splits that may have matches and expose filter attributes for runtime filtering. In case of Delta, it needs to remember a set of files that potentially have rows to update/delete.
- If the data source supports runtime filtering, Spark will assign a filtering subquery that will be executed at runtime based on the SQL command. The same scan object will be used in the filtering subquery and in the row-level operation. In case of Delta, the runtime filtering will be done using the `_file_name` metadata column.
- Spark executes the filtering subquery via the existing runtime filtering mechanism, collects unique values for the filtering attributes and passes them into the scan. In case of Delta, the scan will receive a set of files that have matches.
- The data source uses the passed unique values to filter out cached group IDs or tasks and reports back input splits for groups that definitely have matches. In case of Delta, it can filter out previously cached files and include tasks only for files that have matches.
- Writes have access to scans, so each scan can report a list of affected group IDs (which was cached). It can be in any format: strings, case classes, etc. Whatever a particular data source needs.

Reply:
I believe the current behavior in Delta pretty much matches the above algorithm.

Before considering alternative designs, I wanted to make sure we agree the proposed APIs would work for Delta.
Are we on the same page, @cloud-fan?

Reply:
Yea we agree on the algorithm, the question is how the affected "groups" are determined and passed around. BTW, Delta does want to support conditions with subqueries, and we want to make sure the newly proposed API here can work with subqueries: https://github.com/delta-io/delta/issues/826

> Spark executes the filtering subquery via the existing runtime filtering mechanism, collects unique values for the filtering attributes ...

This is the key problem. How does Spark know how to get the file name column? There is no DS v2 API for Delta to tell Spark: hey you can select `_file_name` column to get the "group id". Are we going to add an implicit assumption that, the `Scan` will add an extra column silently like my proposal?

Reply:
Yeah, the API would definitely work with subqueries. We won't even have to do much to support them. The example above contained a subquery and was rewritten correctly.

```
DELETE FROM t WHERE id IN (SELECT * FROM deleted_id)
```

```
== Optimized Logical Plan ==
ReplaceData RelationV2[id#66, dep#67] t
+- Project [id#66, dep#67]
   +- Filter NOT (exists#86 <=> true)
      +- Join ExistenceJoin(exists#86), (id#66 = value#19)
         :- Project [id#66, dep#67]
         :  +- Filter dynamicpruning#85 [_file_name#70]
         :     :  +- Project [_file_name#83]
         :     :     +- Join LeftSemi, (id#81 = value#72)
         :     :        :- Project [id#81, _file_name#83]
         :     :        :  +- RelationV2[id#81, dep#82, _file_name#83] t
         :     :        +- LocalRelation [value#72]
         :     +- RelationV2[id#66, dep#67, _file_name#70] t
         +- LocalRelation [value#19]
```

> How does Spark know how to get the file name column? There is no DS v2 API for Delta to tell Spark: hey you can select _file_name column to get the "group id". Are we going to add an implicit assumption that, the Scan will add an extra column silently like my proposal? If yes, I think this solves the problem and is actually very similar to my proposal with the difference that we use the existing hidden column and runtime filter API to report "group id" from the source and use it to determine the final affected "groups".

Delta's `Scan` for row-level operations can inherit `SupportsRuntimeFiltering` and report `_file_name` as a filtering attribute via `SupportsRuntimeFiltering$filterAttributes`. The `_file_name` should be a metadata column that can be projected via `RowLevelOperation$requiredMetadataAttributes`. That's what I propose as a way to report which column to use to filter out group IDs. Both metadata columns and runtime filtering is already supported for DS V2.